posts/inside-rust/2023-05-02-stabilizing-async-fn-in-trait.md

```diff
@@ -0,0 +1,197 @@
+
+[RFC 3185]: https://rust-lang.github.io/rfcs/3185-static-async-fn-in-trait.html
+
+
+## MVP Part 2: Send bounds and associated return types
+
+There is one complication that arises when using async functions in traits that doesn't arise with sync functions. Many async runtimes -- notably including the default configurations of [Tokio] and [async-std] -- use a workstealing thread scheduler. This means that futures may move between worker threads dynamically to achieve load balancing. As a result, the future must only capture `Send` data.
+
+[Tokio]: https://tokio.rs/
+
+[async-std]: https://async.rs/
+
+If you author a generic async function that spawns tasks on one of those runtimes, however, you will start to get compilation errors ([playground](XXX)):
+
+```rust
+async fn do_health_check_par(hc: impl HealthCheck) {
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+The problem is that the future returned by `hc.check()` isn't guaranteed to be `Send`. It might access non-Send data. The solution is to add a `Send` bound, but given that this is an async function, it's not obvious how to do that. How do we talk about the future returned by a call to `hc.check()`? Associated return types provide the answer. We can convert the above function to use an explicit type parameter `HC` (instead of `impl HealthCheck`) and then add a new bound, `HC::check(): Send`. This says "the value returned by `HC::check` must be of `Send` type":
+
+```rust
+async fn do_health_check_par<HC>(hc: HC)
+where
+    HC: HealthCheck + Send + 'static,
+    HC::check(): Send, // <-- associated return type
+{
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+Of course, it's kind of unfortunate that we had to rewrite from taking an `impl HealthCheck` to an explicit `HC` type parameter in order to use this notation. RFC #2289, "associated type bounds", introduced a compact notation to address this problem. That RFC is not part of this MVP, but if it were stabilized, then one could simply write:
+
+```rust
+async fn do_health_check_par(hc: impl HealthCheck<check(): Send> + Send + 'static) {
+    //                                            -------------
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+In our [previous post][pp], we [hypothesized](https://blog.rust-lang.org/inside-rust/2022/11/17/async-fn-in-trait-nightly.html#hypothesis-this-is-uncommon) that this problem might not occur often in practice. However, our case studies found that it comes up quite frequently, and so we decided that a solution is needed. We explored a number of solutions and concluded that associated return types are the most practical.
```

Comment:
I don't agree with the conclusion of being "deeply unpleasant to both read and write". The case studies do not support that. The only critical text I could find was in the Microsoft study:

>  Once we did this, we needed to add `save(..): Send` and `restore(..): Send` bounds to the places where the Component trait was used as a DynComponent. There were only two methods we needed to bound, but this was still mildly annoying.

> Return type notation looks like it would work, but even in our limited experience with it we have run into ergonomics issues.

The syntax has become more ergonomic since the studies by removing the need for `..`.

We did discuss adding a proc macro for the cases where you need multiple `Send` bounds for a given trait, to make that more convenient to write, and it might be good to mention that in this post. (Quite possibly for doing the `DynTrait` workaround for you as well.)

But I expect that people will frequently use RTN directly, so I would not frame it as purely a low-level feature. I also don't want to include trait transformers in this post because we are not including it in the MVP. We can mention that we are thinking of ways to make common cases more ergonomic, and that this feature could be seen as a more granular building block for that.

Reply:
I added a mention of an upcoming proc macro alongside the possibility of trait transformers.

Reply:
Hmm, I don't disagree with what you're saying Yosh, but I do feel like you're saying it in stronger terms than I would. I think a proc macro is a good idea to address both dyn dispatch + multiple methods, and all things being equal I think it'd be good for the async WG to ship it (though it'd probably be ok to have it come from community too). I think ART (nee RTN) seem like a flexible solution that is fine for a single method but scales poorly if you have many methods, in which case the proc macro will be very desirable. This is not the same as "deeply unpleasant" though and I wouldn't want those words in the post. I do think it's fine though to mention the proc macro as a part of the strategy. 

(Also, I probably shouldn't have volunteered to write code, that was not realistic. We should coordinate with someone else.)

Reply:
```suggestion
In our [previous post][pp], we [hypothesized](https://blog.rust-lang.org/inside-rust/2022/11/17/async-fn-in-trait-nightly.html#hypothesis-this-is-uncommon) that this problem might not occur often in practice. However, our case studies found that it comes up quite frequently, and so we decided that a solution is needed. We explored a number of solutions and concluded that associated return types (ARTs) are a flexible and reasonably ergonomic building block, which makes them a good fit for an MVP. However, we also found that in traits with many methods, ARTs become verbose. As discussed later in the post, we expect to ship a proc-macro that can help workaround this in the short term while we gain experience.
```

Reply:
Maybe something like this?